### PR TITLE
Created a Jira Consumer for Dracon

### PR DESCRIPTION
**Pull request by ValntinDragan**
This is a Jira Consumer that allows you to publish Issues to your organisation's Jira workspace straight from the Dracon results. The consumer can be configured to your workspace's settings.

Please feel free to provide feedback on the technical approach, suggestions for improvements, or questions to clarify.

## Structure of the app

------------
    │/jira_c 
    ├── main.go              <- makes the magic happen
    ├── /config
    ├───── config.yaml     <- for specifying the settings of the Consumer, such as Default Issue Values or where to map the result fields 
    ├───── config.go         <- parses the config.yaml file, returns a Config object
    ├───── types.go          <- struct definition of the Config object
    ├── /jira
    ├───── api.go              <- interacts with the Jira API, authenticates the client and creates the Issues
    ├───── apiutils.go        <- helper methods that serialize inputs into Jira Issue Objects
    ├── /document
    ├───── document.go      <- parses a protobuf result into Document (Dracon result) objects
    ├───── types.go           <- struct definition of the Document (Dracon result) object
    ├─── /utils
    ├───── utils.go             <- Loads all the protobuf results, returns a list of HashMaps containing all results
--------

Documentation on the Go Jira API here: [https://pkg.go.dev/github.com/andygrunwald/go-jira@v1.12.0?tab=doc#IssueFields](https://pkg.go.dev/github.com/andygrunwald/go-jira@v1.12.0?tab=doc#IssueFields)

## Setting up
More information on this is present in the `README.md` file.
The user needs to edit the config.yaml file to specify what fields they wanted filled in their Jira Issues. Only setting up the first 4 defaultValues (project, issueType, components, affectsVersions) is mandatory, the rest are optional. Then, the user needs to set up Environment Variables for their jira credentials (username, api token, url).

## Test running it in a Dracon pipeline
You can quickly test the consumer as part of a Dracon pipeline with the following steps:

1. Clone the repository and do `plz run //consumers/jira_c:image` to start the docker container
2. Grab the container image label
3. Create a `jira-consumer.yaml` file for the pipeline
4. Inside the `jira-consumer.yaml`, add the container label as the image, and add Environment Variables for username, api token and url
5. (optional) You can also specify additional arguments for the consumer such as: `args: ["--raw", "--severityThreshold", "2"]`
6. Add the consumer to your `pipeline.yaml` file and run as usual

## Results on the Dracon pipeline
I have tested the consumer in 5 different Pipelines using the Bandit and Gosec producers. Over 1500 issues have been created (in a testing project), no errors encountered.

Again, feel free to give feedback :) Thanks!

### CODE REVIEW

File: consumers/jira_c/gojira/gojira.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+		log.Fatalf("Unable to contact Jira: %s", err)
+	}
+	return jiraClient
+}
+
+// assembleIssue parses the Dracon message and serializes it into a Jira Issue object
+func (client goJiraClient) assembleIssue(message string) *jira.Issue {
+	// Assign the default CustomField values specified in the configuration
+	customFields := tcontainer.NewMarshalMap()
+	for _, cf := range client.Configs.DefaultValues.CustomFields {
```
Comment (ValntinDragan) [inline consumers/jira_c/gojira/gojira.go:43]: Run this code only once, after the config.yaml is parsed; right now it's being run everytime a new issue is created

File: consumers/jira_c/gojira/gojira_test.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+func TestAuthJiraClient(t *testing.T) {
+	client := authJiraClient("test_user", "test_token", "test_url")
+	assert.NotEmpty(t, client)
+}
+
+func TestAssembleIssue(t *testing.T) {
+	issue := sampleGoJiraClient.assembleIssue(sampleMessage)
+	assert.NotEmpty(t, issue)
```
Comment (VJftw) [inline consumers/jira_c/gojira/gojira_test.go:49]: Do you think it's worth making more assertions about the issue that was returned? i.e. the values of the fields

File: consumers/jira_c/gojira/gojira.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+			Labels:          client.Configs.DefaultValues.IssueFields["labels"],
+			Unknowns:        customFields,
+		},
+	}
+}
+
+// CreateIssue creates a new issue in Jira
+func (client goJiraClient) CreateIssue(message string) error {
+	issue := client.assembleIssue(message)
+
```
Comment (VJftw) [inline consumers/jira_c/gojira/gojira.go:75]: very subjective nit, you don't need to do this:
I find embedded if statements usually quite hard to follow; so this might be a bit more readable  e.g.:
```
issue := client.assembleIssue(message)
if client.DryRunMode {
 log.Printf("Dry run mode. The following issue would have been created: '%s'", issue.Fields.Summary)
 return nil
}

ri, resp, err := client.JiraClient.Issue.Create(issue)
if err != nil {
	body, _ := ioutil.ReadAll(resp.Body)
	log.Printf("Error occurred posting to Jira. Response body:\n%s", body)
	return err
}
log.Printf("Created Jira Issue ID %s. jira_key=%s", ri.ID, string(ri.Key))
return nil
```

File: consumers/jira_c/gojira/jirautils.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+	var draconResult map[string]string
+	err := json.Unmarshal(jBytes, &draconResult)
+	return draconResult, err
+}
+
+// makeCustomField returns the appropriate interface for a jira CustomField given it's type and values
+// :param fieldType: the type of the field in Jira (single-value, multi-value, float)
+// :param values: list of values to be filled in
+// :return the appropriate interface for a CustomField, given the corresponding fieldType and value(s)
+func makeCustomField(fieldType string, values []string) interface{} {
```
Comment (VJftw) [inline consumers/jira_c/gojira/jirautils.go:71]: what's the impact of this returning `nil` if a field type isn't supported? do you think it's worth returning an error if a field isn't supported?

File: consumers/jira_c/gojira/jirautils.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+		affectsVersions = append(affectsVersions, &jira.AffectsVersion{Name: v})
+	}
+	return affectsVersions
+}
+
+// makeDescription creates the description of an issue's enhanced with extra information from the Dracon Result
+// depending on the tool used, diffrent fields are more suitable as description
+func makeDescription(draconResult map[string]string, extras []string) string {
+	desc := "This issue was automatically generated by the Dracon security pipeline.\n\n"
+	switch draconResult["tool_name"] {
```
Comment (VJftw) [inline consumers/jira_c/gojira/jirautils.go:129]: maybe a future feature would be to have this mapping as YAML configuration that you read in; right now it's tying support to only `gosec`, `spotbugs`, `bandit`. Adding more would require changing code and creating a new release etc.

File: consumers/jira_c/gojira/jirautils.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+		}
+		desc = desc + "{code}" + "\n"
+	}
+	return desc
+}
+
+// makeSummary creates the Summary/Title of an issue
+// depending on the tool used, different fields are more suitable as summary
+func makeSummary(draconResult map[string]string) string {
+	switch draconResult["tool_name"] {
```
Comment (VJftw) [inline consumers/jira_c/gojira/jirautils.go:152]: same here: I think YAML configuration that you read in would be a good additional feature

File: third_party/go/BUILD [hunk 7]
```diff
@@ -20,6 +20,65 @@ package(default_visibility = ["PUBLIC"])
...
+
+go_get(
+    name = "jwt-go",
+    get = "github.com/dgrijalva/jwt-go",
+    licences = ["mit"],
+    revision = "06ea1031745cb8b3dab3f6a236daf2b0aa468b7e",
+)
+
+go_get(
+    name = "jwt",
```
Comment (VJftw) [inline third_party/go/BUILD:54]: do we need this? can't see it used anywhere

File: consumers/jira_c/gojira/jirautils.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+	var newConfig config
+	err = yaml.Unmarshal(configFile, &newConfig)
+	if err != nil {
+		log.Fatalf("Unmarshal: %v", err)
+	}
+	return newConfig
+}
+
+// parseDraconMessage returns a hashmap of the parsed dracon message
+func parseDraconMessage(message string) (map[string]string, error) {
```
[the remark this answers is not in the document]
Comment (ValntinDragan) [reply consumers/jira_c/gojira/jirautils.go]: Due to the way the config.yaml file affects the consumer, it makes more sense to represent Dracon Results as a HashMap rather than the Document object. Example:

If the user specifies in config.yaml that they want to append several fields to the description (e.g. 'severity_text', 'cvss', 'target',...) it's easy to just iterate over these keys and write the code as `Description += DraconResult[key]`.
However if you weren't using a HashMap, you wouldn't be able to access the Dracon Result fields by their name and would have to resort to something like `if key == 'cvss': Description += DraconResult.CVSS`, making the code more cluttered.

As a "middle ground" solution, I've improved the code in `utils.go` to always return Dracon Results as HashMaps (whereas before they were returned as strings), cutting out a layer of conversion.

File: consumers/jira_c/gojira/gojira.go (hunk 1 above)
Comment (ValntinDragan) [reply consumers/jira_c/gojira/gojira.go]: Created a `defaultFields` variable that contains all these default fields specified by the user. They only get created once and the code can just reference their values (or deep-copy them in the case of CustomFields).

File: consumers/jira_c/gojira/jirautils.go (hunk 4 above)
Comment (ValntinDragan) [reply consumers/jira_c/gojira/jirautils.go]: I've tested this and the code doesn't crash, it just results in an empty Field. (Exception is if the Field has a set of supported values and `nil` is not one of them -- but this is not something that you would be able to check in the code; createIssue() will throw you this error later)

So I've added a Warning message if the fieldType isn't supported, but the code can still run.

File: consumers/jira_c/gojira/gojira_test.go (hunk 2 above)
Comment (ValntinDragan) [reply consumers/jira_c/gojira/gojira_test.go]: That's true; implemented

File: consumers/jira_c/gojira/gojira.go (hunk 3 above)
Comment (ValntinDragan) [reply consumers/jira_c/gojira/gojira.go]: I like this, definitely more readable. Thank you! :D

File: third_party/go/BUILD (hunk 7 above)
Comment (ValntinDragan) [reply third_party/go/BUILD]: That's true; I had a feeling `jwt-go` (which is needed) depends on `jwt`, but it's not the case. Thank you

File: consumers/jira_c/gojira/jirautils.go (hunk 5 above)
Comment (ValntinDragan) [reply consumers/jira_c/gojira/jirautils.go]: Will work on implementing this in the future

File: consumers/jira_c/gojira/jirautils.go (hunk 6 above)
Comment (ValntinDragan) [reply consumers/jira_c/gojira/jirautils.go]: Will work on implementing this in the future

File: consumers/jira_c/README.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+DRACON_JIRA_URL="domain your jira workspace is hosted on"
+
+
+## Testing locally
+The following command will test that the app and configuration is working correctly.
+`plz test //consumers/jira_c/...`
+
+
+## Running as part of the Dracon Pipeline
+// TO COMPLETE
```
Comment (steakunderscore) [inline consumers/jira_c/README.md:33]: When will this TODO be done? Just making sure it hasn't been missed.

File: consumers/jira_c/README.md [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+The following arguments can be specified:
+```
+   --dryRun              For debugging. Tickets will not be created
```
[the remark this answers is not in the document]
Comment (steakunderscore) [reply consumers/jira_c/README.md:37]: E.g.  `--dryRun` should be `--dry-run`

File: consumers/jira_c/README.md (hunk 9 above)
Comment (ValntinDragan) [reply consumers/jira_c/README.md]: Won't be included anymore, as it doesn't make sense to have it written here but not in the other consumers; I could instead be working on writing the doc tutorial 'Constructing your own pipeline' on [https://github.com/thought-machine/dracon/blob/master/docs/getting-started/tutorials/constructing-your-own-pipeline.md](https://github.com/thought-machine/dracon/blob/master/docs/getting-started/tutorials/constructing-your-own-pipeline.md)

File: consumers/jira_c/utils/utils.go [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/thought-machine/dracon/consumers"
+
+	v1 "api/proto/v1"
+	"consumers/jira_c/types/document"
+)
+
+func severtiyToText(severity v1.Severity) string {
```
[the remark this answers is not in the document]
Comment (ValntinDragan) [reply consumers/jira_c/utils/utils.go]: This has been checked, the assumption in as desired

**Review by VJftw — APPROVED** [review]
LGTM 👍 thanks!

**Review by steakunderscore — APPROVED** [review]
LGTM. Thanks for this @ValntinDragan!
